File: TSP-Project/src/java/exact/BruteForce.java
import java.util.*;

/**
 * Implementação do algoritmo de Força Bruta para TSP
 * Testa todas as permutações possíveis para encontrar a solução ótima
 */
public class BruteForce {
    
    private TSPInstance instance;
    private TSPResult result;
    private long permutationsGenerated;
    
    public TSPResult solve(TSPInstance instance) {
        this.instance = instance;
        this.result = new TSPResult("BRUTE_FORCE_JAVA");
        this.permutationsGenerated = 0;
        
        System.out.println("Iniciando força bruta para " + instance.nCities + " cidades...");
        long expectedPermutations = factorial(instance.nCities - 1);
        System.out.println("Número de permutações a testar: " + expectedPermutations);
        
        long startTime = System.nanoTime();
        
        // Inicializa com primeiro caminho possível
        result.bestCost = Integer.MAX_VALUE;
        result.bestPath = new int[instance.nCities];
        
        // Gera todas as permutações começando da cidade 1 (fixamos cidade 0)
        int[] citiesWithoutFirst = new int[instance.nCities - 1];
        for (int i = 1; i < instance.nCities; i++) {
            citiesWithoutFirst[i - 1] = i;
        }
        
        generatePermutations(citiesWithoutFirst, 0);
        
        long endTime = System.nanoTime();
        result.executionTime = (endTime - startTime) / 1_000_000_000.0;
        result.nodesExplored = permutationsGenerated;
        
        System.out.println("Permutações testadas: " + permutationsGenerated);
        
        return result;
    }
    
    /**
     * Gera todas as permutações usando algoritmo de Heap
     */
    private void generatePermutations(int[] cities, int startIndex) {
        if (startIndex == cities.length - 1) {
            // Testa esta permutação
            testPermutation(cities);
            return;
        }
        
        for (int i = startIndex; i < cities.length; i++) {
            swap(cities, startIndex, i);
            generatePermutations(cities, startIndex + 1);
            swap(cities, startIndex, i); // backtrack
        }
    }
    
    /**
     * Testa uma permutação específica
     */
    private void testPermutation(int[] cities) {
        permutationsGenerated++;
        
        // Constrói caminho completo (começando com cidade 0)
        int[] fullPath = new int[instance.nCities];
        fullPath[0] = 0;
        System.arraycopy(cities, 0, fullPath, 1, cities.length);
        
        // Calcula custo
        int cost = instance.calculatePathCost(fullPath);
        
        // Atualiza melhor solução se necessário
        if (cost < result.bestCost) {
            result.bestCost = cost;
            System.arraycopy(fullPath, 0, result.bestPath, 0, fullPath.length);
            
            // Mostra progresso a cada melhoria
            System.out.printf("Nova melhor solução encontrada: %d (permutação %d)%n", 
                             cost, permutationsGenerated);
        }
        
        // Mostra progresso a cada 1 milhão de permutações
        if (permutationsGenerated % 1_000_000 == 0) {
            System.out.printf("Progresso: %d permutações testadas...%n", permutationsGenerated);
        }
    }
    
    /**
     * Troca dois elementos no array
     */
    private void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    
    /**
     * Calcula fatorial de um número
     */
    private long factorial(int n) {
        if (n <= 1) return 1;
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }
    
    /**
     * Método para testar a classe isoladamente
     */
    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Uso: java BruteForce <arquivo_tsp>");
            return;
        }
        
        try {
            TSPInstance instance = new TSPInstance(args[0]);
            BruteForce bf = new BruteForce();
            TSPResult result = bf.solve(instance);
            result.printResults(instance);
        } catch (Exception e) {
            System.err.println("Erro: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: src/c/exact/brute_force.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <time.h>

#define MAX_CITIES 50

// Declaração da função factorial
long long factorial(int n);

typedef struct {
    int **matrix;
    int n_cities;
    int *best_path;
    int best_cost;
    double execution_time;
} TSPData;

// Função para ler a matriz de adjacência do arquivo
TSPData* read_tsp_file(const char* filename) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        printf("Erro ao abrir arquivo %s\n", filename);
        return NULL;
    }
    
    TSPData *data = malloc(sizeof(TSPData));
    
    // Conta o número de cidades lendo a primeira linha
    char line[10000];
    if (fgets(line, sizeof(line), file) == NULL) {
        printf("Erro ao ler primeira linha\n");
        fclose(file);
        free(data);
        return NULL;
    }
    int count = 0;
    char *token = strtok(line, " \t\n");
    while (token != NULL) {
        count++;
        token = strtok(NULL, " \t\n");
    }
    
    data->n_cities = count;
    data->best_cost = INT_MAX;
    
    // Aloca memória para a matriz
    data->matrix = malloc(data->n_cities * sizeof(int*));
    for (int i = 0; i < data->n_cities; i++) {
        data->matrix[i] = malloc(data->n_cities * sizeof(int));
    }
    
    data->best_path = malloc(data->n_cities * sizeof(int));
    
    // Volta ao início do arquivo
    rewind(file);
    
    // Lê a matriz
    for (int i = 0; i < data->n_cities; i++) {
        for (int j = 0; j < data->n_cities; j++) {
            if (fscanf(file, "%d", &data->matrix[i][j]) != 1) {
                printf("Erro ao ler matriz na posição [%d][%d]\n", i, j);
                fclose(file);
                // Libera memória
                for (int k = 0; k <= i; k++) {
                    free(data->matrix[k]);
                }
                free(data->matrix);
                free(data->best_path);
                free(data);
                return NULL;
            }
        }
    }
    
    fclose(file);
    return data;
}

int calculate_path_cost(TSPData *data, int *path) {
    int total_cost = 0;
    for (int i = 0; i < data->n_cities - 1; i++) {
        total_cost += data->matrix[path[i]][path[i + 1]];
    }
    total_cost += data->matrix[path[data->n_cities - 1]][path[0]];
    return total_cost;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// Algoritmo de Heap para gerar permutações
void generate_permutations(TSPData *data, int *path, int size, int start_index) {
    if (start_index == size) {
        // Reconstrói caminho completo com cidade 0 no início
        int *full_path = malloc(data->n_cities * sizeof(int));
        full_path[0] = 0;
        memcpy(full_path + 1, path, size * sizeof(int));
        
        int cost = calculate_path_cost(data, full_path);
        if (cost < data->best_cost) {
            data->best_cost = cost;
            memcpy(data->best_path, full_path, data->n_cities * sizeof(int));
        }
        
        free(full_path);
        return;
    }
    
    for (int i = start_index; i < size; i++) {
        swap(&path[start_index], &path[i]);
        generate_permutations(data, path, size, start_index + 1);
        swap(&path[start_index], &path[i]); // backtrack
    }
}

// Algoritmo de Força Bruta principal
void solve_tsp_brute_force(TSPData *data) {
    clock_t start_time = clock();
    
    // Fixa a cidade 0 como início
    int *path = malloc(data->n_cities * sizeof(int));
    for (int i = 0; i < data->n_cities; i++) {
        path[i] = i;
    }
    
    printf("Iniciando força bruta para %d cidades...\n", data->n_cities);
    printf("Número de permutações a testar: %lld\n", factorial(data->n_cities - 1));
    
    // Inicializa melhor solução
    data->best_cost = calculate_path_cost(data, path);
    memcpy(data->best_path, path, data->n_cities * sizeof(int));
    
    // Gera permutações começando da cidade 1 (fixa 0)
    if (data->n_cities > 1) {
        generate_permutations(data, path + 1, data->n_cities - 1, 0);
    }
    
    clock_t end_time = clock();
    data->execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    
    free(path);
}

long long factorial(int n) {
    if (n <= 1) return 1;
    long long result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

void print_results(TSPData *data, const char* filename) {
    printf("\n=== RESULTADOS FORÇA BRUTA ===\n");
    printf("Arquivo: %s\n", filename);
    printf("Número de cidades: %d\n", data->n_cities);
    printf("Melhor custo encontrado: %d\n", data->best_cost);
    printf("Tempo de execução: %.6f segundos\n", data->execution_time);
    printf("Melhor caminho: ");
    for (int i = 0; i < data->n_cities; i++) {
        printf("%d ", data->best_path[i]);
    }
    printf("\n");
}

void save_results(TSPData *data, const char* filename, const char* output_file) {
    FILE *file = fopen(output_file, "a");
    if (file) {
        fprintf(file, "%s,%d,%d,%.6f,BRUTE_FORCE_C\n", 
                filename, data->n_cities, data->best_cost, data->execution_time);
        fclose(file);
    }
}

void free_tsp_data(TSPData *data) {
    for (int i = 0; i < data->n_cities; i++) {
        free(data->matrix[i]);
    }
    free(data->matrix);
    free(data->best_path);
    free(data);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Uso: %s <arquivo_tsp>\n", argv[0]);
        return 1;
    }
    
    TSPData *data = read_tsp_file(argv[1]);
    if (!data) {
        return 1;
    }
    
    if (data->n_cities > 12) {
        printf("AVISO: %d cidades pode demorar muito! Continuar? (s/n): ", data->n_cities);
        char response;
        if (scanf(" %c", &response) != 1) {
            printf("Erro ao ler resposta\n");
            free_tsp_data(data);
            return 0;
        }
        if (response != 's' && response != 'S') {
            printf("Execução cancelada.\n");
            free_tsp_data(data);
            return 0;
        }
    }
    
    solve_tsp_brute_force(data);
    print_results(data, argv[1]);
    save_results(data, argv[1], "../../results/exact_results.txt");
    
    free_tsp_data(data);
    return 0;
}